[FSDP] summon full params staticmethod

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75423

Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

rohan-varma committed:
[FSDP] summon full params staticmethod

Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -308,6 +309,29 @@ def test_summon_full_params_equivalence(self, rank0_only, offload_to_cpu):
+        class FSDPContainer(nn.Module):
+            def __init__(self, fsdp_1, fsdp_2):
+                super().__init__()
+                self.fsdp_1 = fsdp_1
```
Minor: Should we include `self.fsdp_3` and in both `model_fsdp` and `model_no_fsdp`, we initialize it as `DeterministicModel(wrap_fsdp=False)`?
Knowing the implementation, this test suffices, but to test a more general non-FSDP root module, the root may have non-FSDP children.

rohan-varma replied:
I see, do you mean add a self.fsdp_3 and `self.fsdp_3 = DeterministicModel(wrap=False)`?

awgu replied:
Yup.

awgu reviewed (approved):
Stamp to unblock 👍🏼

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -349,15 +373,15 @@ def test_reshard_outside_forward_backward_iteration(self, rank0_only, offload_to
         )
         self.assertEqual(0, inner_param._full_param_padded.storage().size())
-        with model.summon_full_params(rank0_only=rank0_only, writeback=not rank0_only, offload_to_cpu=offload_to_cpu):
+        with model.summon_full_params(model, rank0_only=rank0_only, writeback=not rank0_only, offload_to_cpu=offload_to_cpu):
```
I think its actually better practice to call the staticmethod from the class for better readability of the code?

rohan-varma committed:
Update on "[FSDP] summon full params staticmethod"


Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] summon full params staticmethod"


Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] summon full params staticmethod"


Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] summon full params staticmethod"


Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] summon full params staticmethod"


Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

[ghstack-poisoned]
